mlir/lib/IR/AffineExpr.cpp

```diff
@@ -237,6 +239,144 @@ unsigned AffineDimExpr::getPosition() const {
+        symbolicDivide(binaryExpr.getRHS(), symbolPos, divideType));
+  }
+  case AffineExprKind::Mod: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return getAffineBinaryOpExpr(
+        expr.getKind(), symbolicDivide(binaryExpr.getLHS(), symbolPos, 0),
+        symbolicDivide(binaryExpr.getRHS(), symbolPos, 0));
+  }
+  // Dividing any of the operand by the given symbol.
+  case AffineExprKind::Mul: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    if (!isDivisibleBySymbol(binaryExpr.getLHS(), symbolPos, divideType))
+      return binaryExpr.getLHS() *
+             symbolicDivide(binaryExpr.getRHS(), symbolPos, divideType);
+    return symbolicDivide(binaryExpr.getLHS(), symbolPos, divideType) *
+           binaryExpr.getRHS();
+  }
+  // Dividing first operand only by the given symbol.
+  case AffineExprKind::FloorDiv:
+  case AffineExprKind::CeilDiv: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return getAffineBinaryOpExpr(expr.getKind(),
+                                 symbolicDivide(binaryExpr.getLHS(), symbolPos,
+                                                getDivideType(expr.getKind())),
+                                 binaryExpr.getRHS());
+  }
+  }
+  llvm_unreachable("Unknown AffineExpr");
+}
+
+/// Simplify a semi-affine expression by handling modulo, floordiv or ceildiv
+/// by a symbol when the first operand is divisible by that symbol. It can be
+/// applied to any semiaffine expressions. Return expression can be either of
+/// semiaffine or pure affine expression.
+static AffineExpr simplifySemiAffine(AffineExpr expr) {
+  switch (expr.getKind()) {
+  case AffineExprKind::Constant:
+  case AffineExprKind::DimId:
+  case AffineExprKind::SymbolId:
+    return expr;
+  case AffineExprKind::Add:
+  case AffineExprKind::Mul: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    return getAffineBinaryOpExpr(expr.getKind(),
+                                 simplifySemiAffine(binaryExpr.getLHS()),
+                                 simplifySemiAffine(binaryExpr.getRHS()));
+  }
+  // Check if simplification of second operand is a symbol and first operand is
+  // divisible by it. If the operation is modulo it will output constant 0
+  // expression. In case of floordiv and ceildiv, divide first operand by that
+  // symbol. In any other case simplification not possible.
+  case AffineExprKind::FloorDiv:
+  case AffineExprKind::CeilDiv:
+  case AffineExprKind::Mod: {
+    AffineBinaryOpExpr binaryExpr = expr.cast<AffineBinaryOpExpr>();
+    AffineExpr sLHS = simplifySemiAffine(binaryExpr.getLHS());
+    AffineExpr sRHS = simplifySemiAffine(binaryExpr.getRHS());
+    if (!simplifySemiAffine(binaryExpr.getRHS()).isa<AffineSymbolExpr>())
+      return getAffineBinaryOpExpr(expr.getKind(), sLHS, sRHS);
+    AffineSymbolExpr symbolExpr = binaryExpr.getRHS().cast<AffineSymbolExpr>();
```

Comment:
Instead of `isa` + `cast`, you should be using `dyn_cast` here.

```
AffineSymbolExpr symbolExpr = simplifySemiAffine(binaryExpr.getRHS()).dyn_cast<AffineSymbolExpr>();
if (!symbolExpr)
 ...

```